### PR TITLE
add zkjh ops

### PR DESCRIPTION
**Pull request by FatJhon**
add ops of amax，max，min，sum，prod，argmax，detail functions are as follows：
1. amax(tensor, dim, keepdim)  dim as list is not supported
2. max(tensor, dim, keepdim)  max(tensor)
3. min(tensor, dim, keepdim)  max(tensor)
4. prod(tensor) prod(dim)   bfloat16 is not supported
5. sum(tensor) sum(tensor, dim, keepdim)  dim as list is not supported
6. argmax(tensor, dim, keepdim)

### CODE REVIEW

File: tests/flag_gems/op_accu_test.py [hunk 1]
```diff
@@ -895,3 +896,153 @@ def test_accuracy_triu(shape, diagonal, dtype):
...
+def test_accuracy_max_dim(shape, dim, keepdim, dtype):
+
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+
+    ref_out = torch.max(inp, dim=dim, keepdim=keepdim)
+    with flag_gems.use_gems():
+        res_out = torch.max(inp, dim=dim, keepdim=keepdim)
+    ref_out_value, ref_out_index = ref_out
+    res_out_value, res_out_index = res_out
+    allclose_with_dtype(ref_out_index, res_out_index, torch.int64)
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: I think it's okay to assert indexes equal using "=="

File: tests/flag_gems/op_accu_test.py [hunk 2]
```diff
@@ -895,3 +896,153 @@ def test_accuracy_triu(shape, diagonal, dtype):
...
+def test_accuracy_min_dim(shape, dim, keepdim, dtype):
+
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+
+    ref_out = torch.min(inp, dim=dim, keepdim=keepdim)
+    with flag_gems.use_gems():
+        res_out = torch.min(inp, dim=dim, keepdim=keepdim)
+    ref_out_value, ref_out_index = ref_out
+    res_out_value, res_out_index = res_out
+    allclose_with_dtype(ref_out_index, res_out_index, torch.int64)
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: ditto

File: tests/flag_gems/op_accu_test.py [hunk 3]
```diff
@@ -895,3 +896,153 @@ def test_accuracy_triu(shape, diagonal, dtype):
...
+    allclose_with_dtype(ref_out_value, res_out_value, dtype)
+
+
+@pytest.mark.parametrize(
+    "shape",
+    [(16, i * 3) for i in range(1, 20)],
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: it's recommended to test with larger tensor shapes. tolerance could be amplified by argument reduce_dim.

File: tests/flag_gems/op_accu_test.py [hunk 4]
```diff
@@ -895,3 +896,153 @@ def test_accuracy_triu(shape, diagonal, dtype):
...
+    ref_out = torch.sum(inp)
+    with flag_gems.use_gems():
+        res_out = torch.sum(inp)
+
+    allclose_with_dtype(res_out, ref_out, dtype)
+
+
+@pytest.mark.parametrize(
+    "shape",
+    [(3, i * 4) for i in range(1, 20)],
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: ditto

File: tests/flag_gems/op_accu_test.py [hunk 5]
```diff
@@ -895,3 +896,153 @@ def test_accuracy_triu(shape, diagonal, dtype):
...
+)
+@pytest.mark.parametrize("keepdim", [True, False])
+@pytest.mark.parametrize("dim", [0, 1, None])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.float32, torch.bfloat16])
+def test_accuracy_argmax(shape, dim, keepdim, dtype):
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+    ref_out = torch.argmax(inp, dim=dim, keepdim=keepdim)
+    with flag_gems.use_gems():
+        res_out = torch.argmax(inp, dim=dim, keepdim=keepdim)
+    allclose_with_dtype(res_out, ref_out, res_out.dtype)
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: "==" is okay for integers

File: src/flag_gems/prod.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    mid_ptrs = mid + offset
+    mask = offset < MID_SIZE * block_size
+    mid_val = tl.load(mid_ptrs, mask=mask, other=1.0)
+    # reult = mid_val * mid_val
+    prod_val = tl.cumprod(mid_val, axis=0)
+    out_ptr = out + tl.arange(0, BLOCK_MID)
+    tl.store(out_ptr, prod_val.to(mid_val.dtype))
+
+
+def prod(inp, dim=None, *, dtype=None):
```
Comment (StrongSpoon) [inline src/flag_gems/prod.py]: separate prod.dim_int from prod

File: src/flag_gems/min.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    K = inp.numel() // M // N
+
+    inp = inp.contiguous()
+
+    shape_list = list(shape)
+    shape_list[dim] = 1
+    out_value = torch.empty((M, K), dtype=inp.dtype, device=inp.device)
+    out_index = torch.empty((M, K), dtype=torch.int64, device=inp.device)
+    out_value = out_value.reshape(shape_list)
+    out_index = out_index.reshape(shape_list)
```
Comment (StrongSpoon) [inline src/flag_gems/min.py]: there's no need to reshape. you can initialize out_value & out_index as shape_list.

File: src/flag_gems/min.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    K,
+    BLOCK_M: tl.constexpr,
+    BLOCK_N: tl.constexpr,
+):
+    # set offset
+    pid_m = tl.program_id(0)
+    pid_k = tl.program_id(1)
+    m_offset = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
+    n_offset = tl.arange(0, BLOCK_N)
+    offset = m_offset[:, None, None] * N * K + n_offset[None, :, None] * K + pid_k
```
Comment (StrongSpoon) [inline src/flag_gems/min.py]: 2-dim is enough

File: tests/flag_gems/op_accu_test.py (hunk 1 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: tests/flag_gems/op_accu_test.py (hunk 2 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: tests/flag_gems/op_accu_test.py (hunk 5 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: src/flag_gems/prod.py (hunk 6 above)
Comment (FatJhon) [reply src/flag_gems/prod.py]: done

File: src/flag_gems/min.py (hunk 7 above)
Comment (FatJhon) [reply src/flag_gems/min.py]: done

File: src/flag_gems/min.py (hunk 8 above)
Comment (FatJhon) [reply src/flag_gems/min.py]: done

File: tests/flag_gems/op_accu_test.py [hunk 9]
```diff
@@ -7,6 +7,7 @@
     torch.float16: 1e-3,
     torch.float32: 1.3e-6,
     torch.bfloat16: 0.016,
+    torch.int64: 1.3e-6,
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py:9]: delete

File: src/flag_gems/amax.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        M = inp.numel()
+        block_size = triton.next_power_of_2(math.ceil(math.sqrt(M)))
+        mid_size = triton.cdiv(M, block_size)
+        block_mid = triton.next_power_of_2(mid_size)
+        dtype = inp.dtype
+        mid = torch.empty((mid_size,), dtype=dtype, device=inp.device)
+        if dim is None:
+            out = torch.empty([], dtype=dtype, device=inp.device)
+        else:
+            shape = list(inp.shape)
```
Comment (tongxin) [inline src/flag_gems/amax.py:96]: In case dim=[], it looks like `out` suffices to be a scalar tensor.

File: src/flag_gems/amax.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    result_index = tl.max(inp_vals, axis=1)
+
+    out_ptrs = out + offset_index
+    tl.store(out_ptrs, result_index, mask=mask1)
+
+
+def amax(inp, dim=None, keepdim=False):
+    if __debug__:
+        print("GEMS amax")
+    if dim is None or len(dim) == 0:
```
Comment (tongxin) [inline src/flag_gems/amax.py:86]: can `dim` be an integer?

File: src/flag_gems/amax.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        else:
+            shape = list(inp.shape)
+            for i in range(0, inp.dim()):
+                shape[i] = 1
+            out = torch.empty(shape, dtype=dtype, device=inp.device)
+        amax_kernel_1[(mid_size, 1, 1)](inp, mid, M, block_size)
+        amax_kernel_2[(1, 1, 1)](mid, out, mid_size, block_mid)
+        return out
+    else:
+        dim = dim[0]
```
Comment (tongxin) [inline src/flag_gems/amax.py]: if `dim` is not an integer

File: src/flag_gems/amax.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        N = shape[dim]
+        M = math.prod(shape[:dim])
+        K = inp.numel() // M // N
+
+        inp = inp.contiguous()
+
+        shape_list = list(shape)
+        shape_list[dim] = 1
+        out = torch.empty((M, K), dtype=inp.dtype, device=inp.device)
+        out = out.reshape(shape_list)
```
Comment (tongxin) [inline src/flag_gems/amax.py]: Can be merged.

File: src/flag_gems/amax.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        if not keepdim:
+            out = torch.squeeze(out, dim)
+
+        grid = lambda meta: (
+            triton.cdiv(M, meta["BLOCK_M"]),
+            K,
+        )
```
Comment (tongxin) [inline src/flag_gems/amax.py:124]: Due to the grid size limit, there'll be kernel argument error if K is greater than 65535.

File: src/flag_gems/amax.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    key=[
+        "M",
+        "N",
+    ],
+)
+@triton.heuristics(
+    values={"BLOCK_N": lambda args: triton.next_power_of_2(args["N"])},
+)
+@triton.jit
+def amax_kernel(
```
Comment (tongxin) [inline src/flag_gems/amax.py:56]: some general comments: 
when K is not one, there's large room for improvement. we can leave it for future work;)

File: src/flag_gems/amax.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+@libentry()
+@triton.autotune(
+    configs=[
+        triton.Config({"BLOCK_M": 8}, num_warps=8, num_stages=4),
+        triton.Config({"BLOCK_M": 8}, num_warps=8, num_stages=5),
+        triton.Config({"BLOCK_M": 16}, num_warps=8, num_stages=4),
+        triton.Config({"BLOCK_M": 16}, num_warps=8, num_stages=5),
+        triton.Config({"BLOCK_M": 32}, num_warps=8, num_stages=4),
+        triton.Config({"BLOCK_M": 32}, num_warps=8, num_stages=5),
```
Comment (tongxin) [inline src/flag_gems/amax.py:45]: Depending on the shape, a smaller block_m is needed for sufficient thread block level concurrency. Can we tweak this tuning set a little bit?

Comment (tongxin) [reply src/flag_gems/amax.py:45]: Also changing num_warps might help. AFAIK, num_stages doesn't have effect on reductions

File: src/flag_gems/argmax.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    if dim is None:
+        M = inp.numel()
+        if dtype is None:
+            dtype = inp.dtype
+        block_size = triton.next_power_of_2(math.ceil(math.sqrt(M)))
+        mid_size = triton.cdiv(M, block_size)
+        block_mid = triton.next_power_of_2(mid_size)
+
+        mid_value = torch.empty((mid_size,), dtype=dtype, device=inp.device)
+        mid_index = torch.empty((mid_size,), dtype=torch.int64, device=inp.device)
```
Comment (tongxin) [inline src/flag_gems/argmax.py:101]: here we can save index storage if N is smaller than max int32

File: src/flag_gems/amax.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    else:
+        dim = dim[0]
+        assert dim >= -inp.ndim and dim < inp.ndim, "Invalid dim"
+        shape = inp.shape
+        dim = dim % inp.ndim
+        N = shape[dim]
+        M = math.prod(shape[:dim])
+        K = inp.numel() // M // N
+
+        inp = inp.contiguous()
```
Comment (iclementine) [inline src/flag_gems/amax.py:112]: Enforcing a contiguous layout is pragmatic way to deal with arbitary reduction, but in some cases, it may be suboptimal to do so.

For example, reducing a tensor with shape `(M, N)` and stride `(1, M)` at axis 0, is actually a reduction at contiguous dimension, which allows better performance with resonable optimizations. However, rearange the input to a contiguous layout (Shape`(M, N)`, stride`(N, 1)`) would transform the problem into a harder case( reduction at non-contiguous axis).

So maybe we can derive a genral approach to do some layout analysis to a)figure out whether to make the input contiguous and b)figure out a best axis order to achieve better performance. We can leave it for future works.

File: tests/flag_gems/op_accu_test.py [hunk 19]
```diff
@@ -984,6 +1037,124 @@ def test_accuracy_varmean(shape, dim, correction, keepdim, dtype):
...
+)
+@pytest.mark.parametrize("dtype", [torch.float16, torch.float32, torch.bfloat16])
+def test_accuracy_sum(shape, dtype):
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+
+    ref_out = torch.sum(inp.to(torch.float64))
+    with flag_gems.use_gems():
+        res_out = torch.sum(inp)
+
+    allclose_with_dtype(res_out, ref_out, dtype, reduce_dim=math.prod(shape))
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: replace math.prod(shape) with inp.numel(), import math is not necessary

File: tests/flag_gems/op_accu_test.py [hunk 20]
```diff
@@ -984,6 +1037,124 @@ def test_accuracy_varmean(shape, dim, correction, keepdim, dtype):
...
+@pytest.mark.parametrize(
+    "shape",
+    [(4096, i * 64) for i in range(1, 20)],
+)
+# @pytest.mark.parametrize("keepdim", [True, False])
+# @pytest.mark.parametrize("dim", [None])
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float16])
+def test_accuracy_prod(shape, dtype):
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+    ref_out = torch.prod(inp)
```
Comment (StrongSpoon) [inline tests/flag_gems/op_accu_test.py]: I suggest casting input tensor to torch.float64 as reference for all operators

File: src/flag_gems/amax.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    mask = m_offset[:, None] < M and n_offset[None, :] < N
+    inp_ptrs = inp + offset
+    inp_vals = tl.load(inp_ptrs, mask=mask, other=-float("inf"))
+    result_index = tl.max(inp_vals, axis=1)
+
+    out_ptrs = out + offset_index
+    tl.store(out_ptrs, result_index, mask=mask1)
+
+
+def amax(inp, dim=[], keepdim=False):
```
Comment (iclementine) [inline src/flag_gems/amax.py]: Using mutable data as default value (here `[]`)for python function may not be a good  practice.

Comment (FatJhon) [reply src/flag_gems/amax.py]: same to interface of aten amax, default [] of dim

Comment (FatJhon) [reply src/flag_gems/amax.py]: I will modify it for default value of []. thanks

File: tests/flag_gems/op_accu_test.py (hunk 4 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: tests/flag_gems/op_accu_test.py (hunk 3 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: src/flag_gems/amax.py (hunk 12 above)
Comment (StrongSpoon) [reply src/flag_gems/amax.py]: please mark a 'todo' here in case of dim is a list or tuple

File: src/flag_gems/__init__.py [hunk 22]
```diff
@@ -30,8 +30,17 @@
 from .sub import sub
 from .tanh import tanh
 from .triu import triu
+
+from .max import max
+from .min import min
+from .amax import amax
+from .sum import sum
+from .argmax import argmax
+from .prod import prod
```
Comment (StrongSpoon) [inline src/flag_gems/__init__.py]: update

File: tests/flag_gems/op_accu_test.py (hunk 20 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: tests/flag_gems/op_accu_test.py (hunk 19 above)
Comment (FatJhon) [reply tests/flag_gems/op_accu_test.py]: done

File: src/flag_gems/__init__.py (hunk 22 above)
Comment (FatJhon) [reply src/flag_gems/__init__.py]: done
